File: app/src/main/java/quickbeer/android/views/ProgressIndicatorBar.java
/**
 * This file is part of QuickBeer.
 * Copyright (C) 2016 Antti Poikela <antti.poikela@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package quickbeer.android.views;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver.OnPreDrawListener;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.ScaleAnimation;
import android.view.animation.TranslateAnimation;
import android.widget.FrameLayout;

import quickbeer.android.R;
import quickbeer.android.providers.ProgressStatusProvider.ProgressStatus;
import quickbeer.android.providers.ProgressStatusProvider.Status;
import timber.log.Timber;

import static io.reark.reark.utils.Preconditions.checkNotNull;
import static io.reark.reark.utils.Preconditions.get;
import static quickbeer.android.providers.ProgressStatusProvider.Status.INDEFINITE;
import static quickbeer.android.providers.ProgressStatusProvider.Status.LOADING;

public class ProgressIndicatorBar extends FrameLayout {

    private static final int ANIMATION_SCROLL_DURATION = 1200;
    private static final int ANIMATION_PROGRESS_DURATION = 400;
    private static final int ANIMATION_END_SCALE_DURATION = 400;
    private static final int ANIMATION_END_PAUSE_DURATION = 300;
    private static final int ANIMATION_END_FADE_DURATION = 300;

    private final int progressBarWidth = getResources().getDimensionPixelSize(R.dimen.progress_indicator_width);

    private float barScale = 1.0f;

    @Nullable
    private View progressBar;

    @NonNull
    private ProgressStatus currentProgress = new ProgressStatus(Status.IDLE, 0.0f);

    @NonNull
    private ProgressStatus nextProgress = new ProgressStatus(Status.IDLE, 0.0f);

    public ProgressIndicatorBar(Context context) {
        super(context);
        addPreDrawListener();
    }

    public ProgressIndicatorBar(Context context, AttributeSet attrs) {
        super(context, attrs);
        addPreDrawListener();
    }

    public void setProgress(@NonNull ProgressStatus progress) {
        Timber.v("New progress: " + get(progress));

        nextProgress = progress;

        // Only indefinite progress animation can be interrupted
        if (!isUninterruptibleAnimation()) {
            applyNextStatus();
        }
    }

    private void addPreDrawListener() {
        getViewTreeObserver().addOnPreDrawListener(new OnPreDrawListener() {
            @Override
            public boolean onPreDraw() {
                getViewTreeObserver().removeOnPreDrawListener(this);
                initLayout();
                return true;
            }
        });
    }

    private void initLayout() {
        LayoutParams params = new LayoutParams(
                progressBarWidth,
                ViewGroup.LayoutParams.MATCH_PARENT);

        progressBar = new View(getContext());
        progressBar.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.orange));
        progressBar.setLayoutParams(params);
        progressBar.setVisibility(INVISIBLE);

        addView(progressBar);
        applyNextStatus();
    }

    private void applyNextStatus() {
        // Initialisation may not be finished yet
        if (progressBar == null) {
            return;
        }

        // Check if there's new status to apply
        if (statusesEqual(nextProgress, currentProgress)) {
            return;
        }

        currentProgress = nextProgress;

        switch (currentProgress.getStatus()) {
            case INDEFINITE:
                animateScroller();
                break;
            case LOADING:
                animateToProgress(currentProgress.getProgress());
                break;
            case IDLE:
                animateToEnd();
                break;
        }
    }

    private boolean isUninterruptibleAnimation() {
        if (progressBar == null || progressBar.getAnimation() == null) {
            return false;
        }

        if (progressBar.getAnimation().hasEnded()) {
            return false;
        }

        return true;
    }

    private void animateScroller() {
        Timber.v("animateScroller()");
        checkNotNull(progressBar);

        int animEnd = getWidth() - progressBarWidth;

        TranslateAnimation animation = new TranslateAnimation(0, animEnd, 0, 0);
        animation.setDuration(ANIMATION_SCROLL_DURATION);
        animation.setFillAfter(true);
        animation.setRepeatMode(Animation.REVERSE);
        animation.setRepeatCount(Animation.INFINITE);
        animation.setAnimationListener(new Animation.AnimationListener() {
            @Override public void onAnimationStart(Animation animation) {}
            @Override public void onAnimationEnd(Animation animation) {}

            @Override
            public void onAnimationRepeat(Animation animation) {
                if (nextProgress.getStatus() != INDEFINITE) {
                    applyNextStatus();
                }
            }
        });

        progressBar.setVisibility(VISIBLE);
        progressBar.clearAnimation();
        progressBar.startAnimation(animation);
    }

    private void animateToProgress(float progress) {
        Timber.v("animateToProgress(%s)", progress);
        checkNotNull(progressBar);

        float newScale = (getWidth() * progress / progressBarWidth) + 1;

        Timber.v("animate scale %s -> %s", barScale, newScale);

        ScaleAnimation animation = new ScaleAnimation(barScale, newScale,
                progressBar.getHeight(), progressBar.getHeight());
        animation.setDuration(ANIMATION_PROGRESS_DURATION);
        animation.setFillAfter(true);
        animation.setAnimationListener(new Animation.AnimationListener() {
            @Override public void onAnimationStart(Animation animation) {}
            @Override public void onAnimationRepeat(Animation animation) {}

            @Override
            public void onAnimationEnd(Animation animation) {
                barScale = newScale;
                applyNextStatus();
            }
        });

        progressBar.setVisibility(VISIBLE);
        progressBar.clearAnimation();
        progressBar.startAnimation(animation);
    }

    private void animateToEnd() {
        Timber.v("animateToEnd()");
        checkNotNull(progressBar);

        if (progressBar.getAnimation() == null) {
            // Non-active progress bar doesn't need end animation
            return;
        }

        ScaleAnimation animation = new ScaleAnimation(barScale, (getWidth() / (float) progressBarWidth) + 1,
                progressBar.getHeight(), progressBar.getHeight());
        animation.setDuration(ANIMATION_END_SCALE_DURATION);
        animation.setFillAfter(true);
        animation.setAnimationListener(new Animation.AnimationListener() {
            @Override public void onAnimationStart(Animation animation) {}
            @Override public void onAnimationRepeat(Animation animation) {}

            @Override
            public void onAnimationEnd(Animation animation) {
                barScale = 1.0f;
                animateToHidden();
            }
        });

        progressBar.setVisibility(VISIBLE);
        progressBar.clearAnimation();
        progressBar.startAnimation(animation);
    }

    private void animateToHidden() {
        Timber.v("animateToHidden()");
        checkNotNull(progressBar);

        AlphaAnimation animation = new AlphaAnimation(1.0f, 0.0f);
        animation.setStartOffset(ANIMATION_END_PAUSE_DURATION);
        animation.setDuration(ANIMATION_END_FADE_DURATION);
        animation.setFillAfter(true);

        progressBar.clearAnimation();
        progressBar.startAnimation(animation);
    }

    private static boolean statusesEqual(@NonNull ProgressStatus first, @NonNull ProgressStatus second) {
        if (second.equals(first)) {
            return true;
        }

        // Consider progress only if loading, otherwise matching status is enough
        return first.getStatus() == second.getStatus()
               && (first.getStatus() != LOADING || first.equals(second));
    }
}

File: app/src/main/java/quickbeer/android/ui/progress/ProgressIndicatorBar.kt
/**
 * This file is part of QuickBeer.
 * Copyright (C) 2016 Antti Poikela <antti.poikela></antti.poikela>@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http:></http:>//www.gnu.org/licenses/>.
 */
package quickbeer.android.ui.progress

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.view.animation.ScaleAnimation
import android.view.animation.TranslateAnimation
import android.widget.FrameLayout
import androidx.core.content.ContextCompat
import quickbeer.android.R
import timber.log.Timber

class ProgressIndicatorBar @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {

    private val progressBarWidth = resources.getDimensionPixelSize(R.dimen.progress_indicator_width)
    private var barScale = 1.0f
    private var currentProgress = ProgressStatus(Status.IDLE, 0.0f)
    private var nextProgress = ProgressStatus(Status.IDLE, 0.0f)

    private lateinit var progressBar: View

    fun setProgress(progress: ProgressStatus) {
        Timber.v("New progress: $progress")
        nextProgress = progress

        // Only indefinite progress animation can be interrupted
        if (!isUninterruptibleAnimation) {
            applyNextStatus()
        }
    }

    private fun addPreDrawListener() {
        viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
            override fun onPreDraw(): Boolean {
                viewTreeObserver.removeOnPreDrawListener(this)
                initLayout()
                return true
            }
        })
    }

    private fun initLayout() {
        val params = LayoutParams(
            progressBarWidth,
            ViewGroup.LayoutParams.MATCH_PARENT
        )

        progressBar = View(context).apply {
            setBackgroundColor(ContextCompat.getColor(context, R.color.orange))
            layoutParams = params
            visibility = INVISIBLE
        }

        addView(progressBar)
        applyNextStatus()
    }

    private fun applyNextStatus() {
        // Check if there's new status to apply
        if (statusesEqual(nextProgress, currentProgress)) {
            return
        }

        currentProgress = nextProgress

        when (currentProgress.status) {
            Status.INDEFINITE -> animateScroller()
            Status.LOADING -> animateToProgress(currentProgress.progress)
            Status.IDLE -> animateToEnd()
        }
    }

    private val isUninterruptibleAnimation: Boolean
        get() {
            if (progressBar.animation == null) {
                return false
            }
            return !progressBar.animation.hasEnded()
        }

    private fun animateScroller() {
        Timber.v("animateScroller()")
        val animEnd = width - progressBarWidth
        val animation = TranslateAnimation(0F, animEnd.toFloat(), 0F, 0F).apply {
            duration = ANIMATION_SCROLL_DURATION.toLong()
            fillAfter = true
            repeatMode = Animation.REVERSE
            repeatCount = Animation.INFINITE

            setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(animation: Animation) = Unit
                override fun onAnimationEnd(animation: Animation) = Unit
                override fun onAnimationRepeat(animation: Animation) {
                    if (nextProgress.status !== Status.INDEFINITE) {
                        applyNextStatus()
                    }
                }
            })
        }

        progressBar.apply {
            visibility = VISIBLE
            clearAnimation()
            startAnimation(animation)
        }
    }

    private fun animateToProgress(progress: Float) {
        Timber.v("animateToProgress(%s)", progress)
        val newScale = width * progress / progressBarWidth + 1
        val barHeight = progressBar.height.toFloat()

        Timber.v("animate scale %s -> %s", barScale, newScale)
        val animation = ScaleAnimation(barScale, newScale, barHeight, barHeight).apply {
            duration = ANIMATION_PROGRESS_DURATION.toLong()
            fillAfter = true

            setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(animation: Animation) = Unit
                override fun onAnimationRepeat(animation: Animation) = Unit
                override fun onAnimationEnd(animation: Animation) {
                    barScale = newScale
                    applyNextStatus()
                }
            })
        }

        progressBar.apply {
            visibility = VISIBLE
            clearAnimation()
            startAnimation(animation)
        }
    }

    private fun animateToEnd() {
        Timber.v("animateToEnd()")

        if (progressBar.animation == null) {
            // Non-active progress bar doesn't need end animation
            return
        }

        val animation = ScaleAnimation(
            barScale,
            width / progressBarWidth.toFloat() + 1,
            progressBar.height.toFloat(),
            progressBar.height.toFloat()
        ).apply {
            duration = ANIMATION_END_SCALE_DURATION.toLong()
            fillAfter = true
            setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(animation: Animation) = Unit
                override fun onAnimationRepeat(animation: Animation) = Unit
                override fun onAnimationEnd(animation: Animation) {
                    barScale = 1.0f
                    animateToHidden()
                }
            })
        }

        progressBar.visibility = VISIBLE
        progressBar.clearAnimation()
        progressBar.startAnimation(animation)
    }

    private fun animateToHidden() {
        Timber.v("animateToHidden()")

        val animation = AlphaAnimation(1.0f, 0.0f).apply {
            startOffset = ANIMATION_END_PAUSE_DURATION.toLong()
            duration = ANIMATION_END_FADE_DURATION.toLong()
            fillAfter = true
        }

        progressBar.clearAnimation()
        progressBar.startAnimation(animation)
    }

    companion object {
        private const val ANIMATION_SCROLL_DURATION = 1200
        private const val ANIMATION_PROGRESS_DURATION = 400
        private const val ANIMATION_END_SCALE_DURATION = 400
        private const val ANIMATION_END_PAUSE_DURATION = 300
        private const val ANIMATION_END_FADE_DURATION = 300

        private fun statusesEqual(first: ProgressStatus, second: ProgressStatus): Boolean {
            if (second == first) return true

            // Consider progress only if loading, otherwise matching status is enough
            return first.status === second.status &&
                (first.status !== Status.LOADING || first == second)
        }
    }
}
